### PR TITLE
Microservices observability guide

### PR DESCRIPTION
**Pull request by nipunthathsara**
This guide discusses the observability of Ballerina microservices.
Initial pull request.

### CODE REVIEW

File: guide/airline_reservation/airline_reservation_service.bal [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+//}
+//
+//@kubernetes:Deployment {
+//  image:"ballerina.guides.io/airline_reservation_service:v1.0",
+//  name:"ballerina-guides-airline-reservation-service"
+//}
+
+// Service endpoint
+endpoint http:Listener airlineEP {
+    port:9091
```
Comment (hemikak) [inline guide/airline_reservation/airline_reservation_service.bal]: Better to have a space after colon.

File: guide/airline_reservation/airline_reservation_service.bal [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+};
+
+// Airline reservation service
+@http:ServiceConfig {basePath:"/airline"}
+service<http:Service> airlineReservationService bind airlineEP {
+
+    // Resource 'flightQatar', which checks about airline 'Qatar Airways'
+    @http:ResourceConfig {
+        methods:["POST"],
```
Comment (hemikak) [inline guide/airline_reservation/airline_reservation_service.bal]: Better to have a space after colon.

File: guide/airline_reservation/airline_reservation_service.bal [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    log:printDebug("Invoking airlineDBService with parameters - airline : " + airline + ", departureDate : " + departureDate 
+    + ", arrivalDate : " + arrivalDate + ", to : " + to + ", from : " + rom);
+    // Set arguments for the query
+    sql:Parameter p1 = {sqlType:sql:TYPE_VARCHAR, value:airline};
+    sql:Parameter p2 = {sqlType:sql:TYPE_DATE, value:departureDate};
+    sql:Parameter p3 = {sqlType:sql:TYPE_DATE, value:arrivalDate};
+    sql:Parameter p4 = {sqlType:sql:TYPE_VARCHAR, value:to};
+    sql:Parameter p5 = {sqlType:sql:TYPE_VARCHAR, value:rom};
+    // Query to be executed
+    string q = "SELECT * FROM FLIGHTS WHERE airline = ? AND departureDate = ? AND arrivalDate = ? AND dest = ? AND rom = ?";
```
Comment (VijithaEkanayake) [inline guide/airline_reservation/airline_reservation_service.bal]: Better to avoid single letter variables and use meaningful name

File: guide/airline_reservation/airline_reservation_service.bal [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        }
+
+        string arrivalDate = <string> reqPayload.ArrivalDate but {error => ""};
+        string departureDate = <string> reqPayload.DepartureDate but {error => ""};
+        string rom = <string> reqPayload.From but {error => ""};
+        string to = <string> reqPayload.To but {error => ""};
+        string airline = "Qatar";
+
+        // If payload parsing fails, send a "Bad Request" message as the response
+        if (arrivalDate == null || departureDate == null || rom == null || to == null) {
```
Comment (hemikak) [inline guide/airline_reservation/airline_reservation_service.bal]: arrival date would be "" (Empty) if parsing fails. 'null' wont work right ? . Same for other fields ? was this tested ?

File: README.md [hunk 5]
```diff
@@ -1 +1,583 @@
...
+< content-length: 11
+< server: ballerina/0.980.0
+< date: Sun, 5 Aug 2018 00:25:25 +0530
+< 
+* Connection #0 to host localhost left intact
+call failed
+```
+Jaeger traces
+
+![Unavaialbel](images/service-unavailable-trace.png "Service unavaiable")
```
Comment (MaryamZi) [inline README.md]: shall we fix the typo with `Unavaialbel` here and in L367?

File: guide/airline_reservation/airline_reservation_service.bal [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        }
+
+        string arrivalDate = <string> reqPayload.ArrivalDate but {error => ""};
+        string departureDate = <string> reqPayload.DepartureDate but {error => ""};
+        string rom = <string> reqPayload.From but {error => ""};
+        string to = <string> reqPayload.To but {error => ""};
+        string airline = "Asiana";
+
+        // If payload parsing fails, send a "Bad Request" message as the response
+        if (arrivalDate == null || arrivalDate == null || rom == null || to == null) {
```
Comment (VijithaEkanayake) [inline guide/airline_reservation/airline_reservation_service.bal]: Is it correct to use null check here?

File: guide/airline_reservation/airline_reservation_service.bal [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    // Resource 'flightEmirates', which checks about airline 'Emirates'
+    @http:ResourceConfig {
+        methods:["POST"],
+        path:"/emirates",
+        consumes:["application/json"],
+        produces:["application/json"]
+    }
+    flightEmirates (endpoint caller, http:Request request) {
```
Comment (hemikak) [inline guide/airline_reservation/airline_reservation_service.bal]: You seem to have 3 resources with the same functionality. You can use path params ("/{vendor}") and have one resource. Please check.

File: guide/airline_reservation/airline_reservation_service.bal [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    int flightNo;
+    string airline;
+    string arrivalDate;
+    string departureDate;
+    string to;
+    string rom;
+    int price;
+};
+
+function airlineDBService (string airline, string departureDate, string arrivalDate, string to, string rom) returns (json){
```
Comment (hemikak) [inline guide/airline_reservation/airline_reservation_service.bal]: Better if we can add function documentation.

File: guide/airline_reservation/airline_reservation_service.bal [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+        // Response payload
+        log:printDebug("Client response from Emirates : " + flightDetails.toString());
+        response.setJsonPayload(flightDetails);
+        // Send the response to the caller
+        _ = caller->respond(response);
+    }
+}
+
+// Define Flight record type
+type Flight record {
```
Comment (hemikak) [inline guide/airline_reservation/airline_reservation_service.bal]: Better if we can add documentation.

File: guide/airline_reservation/airline_reservation_service.bal (hunk 4 above)
Comment (VijithaEkanayake) [inline guide/airline_reservation/airline_reservation_service.bal]: I think the null check is not the correct methodology to check the existence of the payload attributes

File: guide/airline_reservation/airline_reservation_service.bal [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    int flightNo;
+    string airline;
+    string arrivalDate;
+    string departureDate;
+    string to;
+    string rom;
+    int price;
+};
```
Comment (chamil321) [inline guide/airline_reservation/airline_reservation_service.bal]: Shouldn't the record be a sealed/closed as it doesn't require any extra fields?

File: guide/airline_reservation/airline_reservation_service.bal [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+        log:printDebug("Received at : " + resourcePath);
+        // Try parsing the JSON payload from the request
+        match request.getJsonPayload() {
+            // Valid JSON payload
+            json payload => reqPayload = payload;
+            // NOT a valid JSON payload
+            any => {
+                response.statusCode = 400;
+                response.setJsonPayload({"Message":"Invalid payload - Not a valid JSON payload"});
+                _ = caller->respond(response);
```
Comment (MaryamZi) [inline guide/airline_reservation/airline_reservation_service.bal]: shall we handle the possible errors, maybe with `but`, wherever applicable instead of ignoring them? 

https://github.com/ballerina-platform/ballerina-examples/blob/master/examples/hello-world-service/hello_world_service.bal#L14

File: guide/car_rental/car_rental_service.bal [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+            }
+        }
+
+        string arrivalDate = <string> reqPayload.ArrivalDate but {error => ""};
+        string departureDate = <string> reqPayload.DepartureDate but {error => ""};
+        string vehicleType = <string> reqPayload.VehicleType but {error => ""};
+        string company = "DriveSG";
+
+        // If payload parsing fails, send a "Bad Request" message as the response
+        if (arrivalDate == null || departureDate == null || vehicleType == null) {
```
Comment (hemikak) [inline guide/car_rental/car_rental_service.bal]: Does null work here ?

File: guide/hotel_reservation/hotel_reservation_service.bal [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        // Mock logic
+        // Hotel service doesn't do database calls to retrieve information, response in hard-coded
+        // Details of the hotel
+        json hotelDetails = {
+            "HotelName":"Elizabeth",
+            "FromDate":arrivalDate,
+            "ToDate":departureDate,
+            "DistanceToLocation":2
```
Comment (hemikak) [inline guide/hotel_reservation/hotel_reservation_service.bal]: Space after colon.

File: issue_template.md [hunk 14]
```diff
@@ -0,0 +1,18 @@
+**Description:**
+<!-- Give a brief description of the issue -->
+
```
Comment (hemikak) [inline issue_template.md]: Do we need this file ?

File: pull_request_template.md [hunk 15]
```diff
@@ -0,0 +1,52 @@
+## Purpose
+> Describe the problems, issues, or needs driving this feature/fix and include links to related issues in the following format: Resolves issue1, issue2, etc.
+
```
Comment (hemikak) [inline pull_request_template.md]: Do we need this file ?

File: resources/mysql.sql [hunk 16]
```diff
@@ -0,0 +1,32 @@
+CREATE DATABASE IF NOT EXISTS testdb2;
```
Comment (hemikak) [inline resources/mysql.sql]: License header missing.

File: guide/airline_reservation/airline_reservation_service.bal [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+        json reqPayload;
+
+        string resourcePath = "/airline/qatarAirways";
+        log:printDebug("Received at : " + resourcePath);
+        // Try parsing the JSON payload from the request
+        match request.getJsonPayload() {
+            // Valid JSON payload
+            json payload => reqPayload = payload;
+            // NOT a valid JSON payload
+            any => {
```
Comment (MaryamZi) [inline guide/airline_reservation/airline_reservation_service.bal]: `request.getJsonPayload()` returns json|error right? shall we have the match clause for `error` instead of `any`?

File: guide/car_rental/car_rental_service.bal [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+        json carDetails = untaint carDBService(company, departureDate, arrivalDate, vehicleType);
+        // Response payload
+        log:printDebug("Client response : " + carDetails.toString());
+        response.setJsonPayload(carDetails);
+        // Send the response to the caller
+        _ = caller->respond(response);
+    }
+}
+
+// Define car record type
```
Comment (chamil321) [inline guide/car_rental/car_rental_service.bal]: Shall we document record fields?

File: guide/car_rental/car_rental_service.bal [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+type Car record {
+    string company;
+    string arrivalDate;
+    string departureDate;
+    string vehicleType;
+    int price;
+};
```
Comment (chamil321) [inline guide/car_rental/car_rental_service.bal]: Shouldn't the record be a sealed/closed as it doesn't require any extra fields?

File: guide/airline_reservation/airline_reservation_service.bal [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+import ballerina/log;
+import ballerina/mysql;
+import ballerina/observe;
+import ballerina/runtime;
+import ballerina/sql;
+//import ballerinax/docker;
```
Comment (sinthuja) [inline guide/airline_reservation/airline_reservation_service.bal]: Why is the commented code here? If it's not required, let's just remove it. Else please include a comment to describe the information about the commented code, and when we need to uncomment it.

File: README.md [hunk 21]
```diff
@@ -1 +1,583 @@
...
+
+Travel Agency service communicates with multiple services. Airline reservation service has multiple resources for three different airlines, whereas Hotel reservation and Car rental services only have a single resource. The travel agency service checks all airline resources in parallel to select the cheapest available. Each associated service then queries their database for the availability and information of their services. Only Hotel resevation service does not have a associated database. Rather the response is hard-coded to demonstrate the difference in traces.
+
+![Ballerina Observability Sample](/images/ballerina-observability-sample.png)
+
+## Prerequisites
+ 
+- [Ballerina Distribution](https://ballerina.io/learn/getting-started/)
+- A Text Editor or an IDE
+- [Docker](https://docs.docker.com/install/linux/docker-ce/ubuntu/)
```
Comment (hemikak) [inline README.md]: Docker in not required for running this ?

File: README.md [hunk 22]
```diff
@@ -1 +1,583 @@
...
+
+Consider a real-world use case of a travel agency that arranges complete tours for users. A tour package includes airline ticket reservation, hotel room reservation and car rental. Therefore, the travel agency service requires communicating with other necessary back-ends. Those back-ends query their databases to retrieve information of availability of their services.
+
+Travel Agency service communicates with multiple services. Airline reservation service has multiple resources for three different airlines, whereas Hotel reservation and Car rental services only have a single resource. The travel agency service checks all airline resources in parallel to select the cheapest available. Each associated service then queries their database for the availability and information of their services. Only Hotel resevation service does not have a associated database. Rather the response is hard-coded to demonstrate the difference in traces.
+
+![Ballerina Observability Sample](/images/ballerina-observability-sample.png)
+
+## Prerequisites
+ 
+- [Ballerina Distribution](https://ballerina.io/learn/getting-started/)
```
Comment (hemikak) [inline README.md]: Ballerina runtime would also work ?

File: .travis.yml [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+script:
+# Declare STRING variable
+- STRING="Downloading ballerina..."
```
Comment (Shan1024) [inline .travis.yml]: Can we update the Travis config to the format of - https://github.com/ballerina-guides/playground-post-service/blob/master/.travis.yml#L15

**Review by pubudu91 — CHANGES_REQUESTED** [review]

File: resources/mysql.sql [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+  `departureDate` DATE,
+  `vehicleType` varchar(100),
+  `price` int(10),
+  PRIMARY KEY (company)
+) ENGINE=InnoDB DEFAULT CHARSET=latin1;
+
+INSERT INTO `CARS` (`company`, `arrivalDate`, `departureDate`, `vehicleType`, `price`) VALUES
+('DriveSG', '2007-11-06', '2007-11-06', 'Car', 10),
+('DreamCar', '2007-11-06', '2007-11-06', 'Car', 20),
+('Sixt', '2007-11-06', '2007-11-06', 'Car', 30);
```
Comment (Shan1024) [inline resources/mysql.sql]: Missin newline?

File: guide/travel_agency/travel_agency_service_parallel.bal [hunk 25]
```diff
@@ -0,0 +1,265 @@
...
+            }
+
+            // Worker to communicate with airline 'Emirates'
+            worker emiratesWorker {
+                http:Request outReq;
+                // Out request payload
+                outReq.setJsonPayload(untaint flightPayload);
+                log:printDebug("Sending request to : /emirates");
+                // Send a POST request to 'Emirates' and get the results
+                http:Response respWorkerEmirates = check airlineEP -> post("/emirates", outReq);
```
Comment (Bhashinee) [inline guide/travel_agency/travel_agency_service_parallel.bal]: Remove spaces before and after '->'. In line 172 also.

File: guide/hotel_reservation/tests/hotel_reservation_service_test.bal [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+
+// Client endpoint
+endpoint http:Client clientEP {
+    url:"http://localhost:9092/hotel"
+};
+
+// Function to test resource 'elizabeth'
+@test:Config
+function testResourceElizabeth () {
+    // Initialize the empty http requests and responses
```
Comment (arukshani) [inline guide/hotel_reservation/tests/hotel_reservation_service_test.bal]: Here you only create a new request, not a response. This comment should be corrected in all the places it's been used this way.

File: README.md [hunk 27]
```diff
@@ -1 +1,583 @@
...
+- Identifying the bottleneck
+- Identifying the unavailble services
+- Determining database call duration and bottleneck queries
+- Function execution time (user-defined observability)
+
+Consider a real-world use case of a travel agency that arranges complete tours for users. A tour package includes airline ticket reservation, hotel room reservation and car rental. Therefore, the travel agency service requires communicating with other necessary back-ends. Those back-ends query their databases to retrieve information of availability of their services.
+
+Travel Agency service communicates with multiple services. Airline reservation service has multiple resources for three different airlines, whereas Hotel reservation and Car rental services only have a single resource. The travel agency service checks all airline resources in parallel to select the cheapest available. Each associated service then queries their database for the availability and information of their services. Only Hotel resevation service does not have a associated database. Rather the response is hard-coded to demonstrate the difference in traces.
```
Comment (pubudu91) [inline README.md]: Hotel reservation -> hotel reservation
Car rental -> car rental

Shall we rephrase the last two sentences?

File: README.md (hunk 22 above)
Comment (pubudu91) [reply README.md]: The runtime is not available on the website

File: README.md [hunk 28]
```diff
@@ -1 +1,583 @@
...
+ 
+- [Ballerina Distribution](https://ballerina.io/learn/getting-started/)
+- A Text Editor or an IDE
+- [Docker](https://docs.docker.com/install/linux/docker-ce/ubuntu/)
+- [Prometheus](https://ballerina.io/learn/how-to-observe-ballerina-code/#prometheus)
+- [Grafana](https://ballerina.io/learn/how-to-observe-ballerina-code/#grafana)
+- [Jaeger](https://ballerina.io/learn/how-to-observe-ballerina-code/#jaeger-server)
+
+### Optional requirements
+- Ballerina IDE plugins ([IntelliJ IDEA](https://plugins.jetbrains.com/plugin/9520-ballerina), [VSCode](https://marketplace.visualstudio.com/items?itemName=WSO2.Ballerina), [Atom](https://atom.io/packages/language-ballerina))
```
Comment (pubudu91) [inline README.md]: Shall we remove Atom from this? We don't maintain this plugin

File: README.md [hunk 29]
```diff
@@ -1 +1,583 @@
...
+## Testing the services
+
+### Setting up the environment  
+
+Please refer to the [Prerequisites](#prerequisites) section for information on setting up the required tools and environment. Access the below URLs and make sure you have Prometheus, Grafana and Jaeger servers up and running.
+- Prometheus - [http://localhost:19090](http://localhost:19090)
+- Grafana - [http://localhost:3000](http://localhost:3000)
+- Jaeger - [http://localhost:16686](http://localhost:16686)
+
+Create the database using the script provided in resources directory. This datbase is being used by the services to retrieve data.
```
Comment (pubudu91) [inline README.md]: in -> in the 
This datbase -> This database

Shall we rephrase the second sentence to something like "This will be the database used by the services."

File: README.md [hunk 30]
```diff
@@ -1 +1,583 @@
...
+mysql -u <username> -p
+source resources/mysql.sql;
+```
+
+### Invoking the service
+
+- Navigate to `microservices-observability/guide` and run the following commands in separate terminals to start all four HTTP services. This will start the `Airline Reservation`, `Hotel Reservation`, `Car Rental` and `Travel Agency` services in ports 9091, 9092, 9093 and 9090 respectively.
+
+```bash
+   $ ballerina run travel_agency/
```
Comment (pubudu91) [inline README.md]: Shall we get rid of the trailing `/`? It's not part of the package name. Also, I'm not sure of the use of `_` in the package name.

Comment (pubudu91) [reply README.md]: Fix the above in other relevant places as well

File: guide/ballerina.conf [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+[b7a]
+
+[b7a.observability]
```
Comment (pubudu91) [inline guide/ballerina.conf]: This is also not required IIRC

File: README.md [hunk 32]
```diff
@@ -1 +1,583 @@
...
+```bash 
+   $ ballerina run -e b7a.observability.metrics.prometheus.port=9797 airline_reservation/
+```
+```bash
+   $ ballerina run -e b7a.observability.metrics.prometheus.port=9798 hotel_reservation/
+```
+```bash
+   $ ballerina run -e b7a.observability.metrics.prometheus.port=9799 car_rental/
+```
+Note that we are overridng the prometheus port by passing the parameter (`-e b7a.observability.metrics.prometheus.port=xxxx`) for all the services, except the travel_agency service. Travel agency service will publish the metrics data on the default port defined in the `ballerina.conf` file (9796). This will avoid any conflicts on port being already occupied.
```
Comment (pubudu91) [inline README.md]: prometheus -> Prometheus

Remove the parenthesis around the config param

Instead of referring to the travel agency service by its package name, shall we use the actual name of the service or maybe Travel Agency service? Whichever is used, need to use it consistently. Immediately, in the second sentence it says "Travel agency service".

Comment (pubudu91) [reply README.md]: data on the -> data to the

File: README.md [hunk 33]
```diff
@@ -1 +1,583 @@
...
+- Invoke the Travel Agency service by sending a POST request to arrange a tour.
+
+```bash
+   curl -v -X POST -d \
+   '{"ArrivalDate":"2007-11-06", "DepartureDate":"2007-11-06", "From":"CMB", "To":"DXB",
```
Comment (pubudu91) [inline README.md]: Maybe use a more current date?

File: README.md [hunk 34]
```diff
@@ -1 +1,583 @@
...
+            "HotelName":"Elizabeth","FromDate":"2007-11-06","ToDate":"2007-11-06","DistanceToLocation":2
+        },
+        "Vehicle":{
+            "company":"Sixt","arrivalDate":"2007-11-06+05:30","departureDate":"2007-11-06+05:30",
+            "vehicleType":"Car","price":30
+        }
+    }
+```
+## Observability 
+Ballerina is by default observable. Meaning, you can easily observe your services, resources, etc. Observability configurations are defined in the `ballerina.conf` file. A sample configuration file can be found in `microservices-observability/guide/` directory.
```
Comment (pubudu91) [inline README.md]: Shall we rephrase this to something like the one in https://github.com/ballerina-guides/restful-service/blob/master/README.md#observability ?

File: README.md [hunk 35]
```diff
@@ -1 +1,583 @@
...
+To start the Ballerina services using the configuration file, run the following command
+```
+   $ ballerina run --config travel_agency/ballerina.conf <package_name>
+```
+When you execute the ```ballerina run``` command from the ```microservices-observability/guide/``` location (where the `ballerina.conf` file is), the configuration file will get applied to every service that you are starting. Therefore, you don't need to pass the `--config` argument explicitly.
+
+### Tracing 
+You can monitor Ballerina services using in-built tracing capabilities of Ballerina. We'll use [Jaeger](https://github.com/jaegertracing/jaeger) as the distributed tracing system.
+Follow the below steps to use tracing with Ballerina.
+- Add below configurations to the `ballerina.conf` file
```
Comment (pubudu91) [inline README.md]: Add below -> Add the following

File: README.md [hunk 36]
```diff
@@ -1 +1,583 @@
...
+   $ docker run -d -p5775:5775/udp -p6831:6831/udp -p6832:6832/udp -p5778:5778 \
+   -p16686:16686 p14268:14268 jaegertracing/all-in-one:latest
+```
+
+- Observe the tracers using Jaeger UI using following URL after doing some service invocations on Travel Agency service.
+
+   [http://localhost:16686](http://localhost:16686)
+
+
+- You should see the Jaeger UI as follows
```
Comment (pubudu91) [inline README.md]: The Jaeger UI should look similar to the following:

File: README.md [hunk 37]
```diff
@@ -1 +1,583 @@
...
+reporter.max.buffer.spans=1000
+```
+
+- Run Jaeger Docker image using the following command. (If you haven't already started Jaeger as instructed in the [prerequisites](#prerequisites) section)
+```bash
+   $ docker run -d -p5775:5775/udp -p6831:6831/udp -p6832:6832/udp -p5778:5778 \
+   -p16686:16686 p14268:14268 jaegertracing/all-in-one:latest
+```
+
+- Observe the tracers using Jaeger UI using following URL after doing some service invocations on Travel Agency service.
```
Comment (pubudu91) [inline README.md]: Invoke the Travel Agency service several times and observe the tracers in the Jaeger UI using the following URL

**Review by pubudu91 — CHANGES_REQUESTED** [review]

File: guide/airline_reservation/airline_reservation_service.bal [hunk 38]
```diff
@@ -0,0 +1,268 @@
...
+    foreach i in flights {
+        flight.flightNo = i.flightNo;
+        flight.airline = i.airline;
+        flight.departureDate = i.departureDate;
+        flight.arrivalDate = i.arrivalDate;
+        flight.to = i.to;
+        flight.rom = i.rom;
+        flight.price = i.price;
+    }
+    log:printDebug("airlineDBService response : " );
```
Comment (Anoukh) [inline guide/airline_reservation/airline_reservation_service.bal]: Have we missed to add the flight details here?

File: issue_template.md (hunk 14 above)
Comment (nipunthathsara) [reply issue_template.md]: It is there in other Ballerina guides

File: pull_request_template.md (hunk 15 above)
Comment (nipunthathsara) [reply pull_request_template.md]: It is there in other Ballerina guides.

File: README.md (hunk 21 above)
Comment (nipunthathsara) [reply README.md]: Observability tools used in this guide requires Docker. Hence, must have.

File: README.md [hunk 39]
```diff
@@ -1 +1,583 @@
...
\ No newline at end of file
+[![Build Status](https://travis-ci.org/ballerina-guides/microservices-observability.svg?branch=master)](https://travis-ci.org/ballerina-guides/microservices-observability)
+
+# Observing Ballerina Microservices
+
+Observability of services is useful to retrieve internal information for easier troubleshooting and system status analysis. Ballerina allows to observe its services using Monitoring, Logging and Distributed tracing methods, with the use of other external tools.
+
+Ballerina observability can be categorized into two types as,
+- Default observability - Each Ballerina service can expose its internal information just by enabling the built-in observability. (Eg : Network call duration)
+- User-defined observability - More specific data can be retrieved by using user-defined observability. (Eg : Function execution time) This requires the explicit import of the package `ballerina/observe`.
```
[the remark this answers is not in the document]
Comment (nipunthathsara) [reply README.md]: Explained with an example later in the Usages of Observalbility section.

File: README.md [hunk 40]
```diff
@@ -1 +1,583 @@
-# microservices-observability
\ No newline at end of file
+[![Build Status](https://travis-ci.org/ballerina-guides/microservices-observability.svg?branch=master)](https://travis-ci.org/ballerina-guides/microservices-observability)
+
+# Observing Ballerina Microservices
+
+Observability of services is useful to retrieve internal information for easier troubleshooting and system status analysis. Ballerina allows to observe its services using Monitoring, Logging and Distributed tracing methods, with the use of other external tools.
```
Comment (Nashaath) [inline README.md]: Minor improvement:
"Observability of services is useful to retrieve internal information for easier troubleshooting and system status analysis. Ballerina allows you to observe its services via monitoring, logging and distributed tracing methods that use external tools."

File: README.md (hunk 28 above)
Comment (nipunthathsara) [reply README.md]: [Regarding the runtime review](https://github.com/ballerina-guides/microservices-observability/pull/1#discussion_r217954478), is runtime publicly available in anywhere? All I see is the BDK (or whatever it is called) not the runtime separately.

File: guide/airline_reservation/airline_reservation_service.bal (hunk 20 above)
Comment (nipunthathsara) [reply guide/airline_reservation/airline_reservation_service.bal]: Required only when deploying the services in Docker. Necessary instructions are provided in the Readme file under **Deploying on Docker** section.

File: guide/airline_reservation/airline_reservation_service.bal (hunk 7 above)
Comment (nipunthathsara) [reply guide/airline_reservation/airline_reservation_service.bal]: Adopting this will affect on Jaeger visualizations which conflicts with the intention of this guide.

**Review by kasun04 — CHANGES_REQUESTED** [review]
- Elaborate end to end tracing capabilities across multiple services. How to trace a given message ID across multiple services. This applies to distributed logs of all services too. 
- Docker, K8s has to be updated with observability tools as well. (how to run them in the containerized environment and the respective ballerina configs has to change according to the deployment) 
- Verify with the observability team on whether we have used the most appropriate tool for a given observability aspect.

**Comment by anupama-pathirage** [comment]
Closing this for now as the requested changes are not addressed and we are not actively maintaining ballerina guides.
